### PR TITLE
Feature/rewards v2

### PR DESCRIPTION
**Pull request by stanieltron**
rewards v2 implementation,
migration fn

### CODE REVIEW

**Review by mateuszaaa — CHANGES_REQUESTED** [review]
please verify your commits and rework git history

File: pallets/issuance/src/lib.rs [hunk 1]
```diff
@@ -302,38 +324,61 @@ pub mod pallet {
...
-	/// should be only used for testing purposes
 	fn initialize() {
 		IsTGEFinalized::<T>::put(true);
 		Self::do_init_issuance_config().unwrap();
 	}
 
-	fn compute_issuance(n: u32) {
-		let _ = Pallet::<T>::calculate_and_store_round_issuance(n);
-		let _ = Pallet::<T>::clear_round_issuance_history(n);
+	fn compute_issuance(n: u32) -> DispatchResult {
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/issuance/src/lib.rs:337]: @stanieltron Why was this changed?

File: pallets/xyk/src/lib.rs [hunk 2]
```diff
@@ -790,316 +889,247 @@ impl<T: Config> Pallet<T> {
...
+		let mut current_rewards = 0;
 
-	pub fn calculate_rewards_amount(
-		user: AccountIdOf<T>,
-		liquidity_asset_id: TokenId,
-	) -> Result<Balance, DispatchError> {
-		ensure!(Self::is_promoted_pool(liquidity_asset_id), Error::<T>::NotAPromotedPool);
+		if liquidity_assets_amount != 0 {
+			let pool_rewards_ratio_current =
+				<T as Config>::PoolPromoteApi::get_pool_rewards_v2(liquidity_asset_id).unwrap();
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:904]: please get rid of all unwraps

File: pallets/xyk/src/lib.rs [hunk 3]
```diff
@@ -1112,112 +1142,82 @@ impl<T: Config> Pallet<T> {
...
-		) = Self::calculate_liquidity_checkpoint(user.clone(), liquidity_asset_id, 0 as u128)?;
+		let current_time: u32 = (<frame_system::Pallet<T>>::block_number().saturated_into::<u32>() +
+			1) / T::RewardsDistributionPeriod::get();
 
-		let liquidity_assets_amount: Balance =
-			LiquidityMiningActiveUser::<T>::get((&user, &liquidity_asset_id));
-		let activated_liquidity_pool: Balance =
-			LiquidityMiningActivePool::<T>::get(&liquidity_asset_id);
+		let mut pool_ratio_current =
+			<T as Config>::PoolPromoteApi::get_pool_rewards_v2(liquidity_asset_id).unwrap();
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:1154]: get rid of unwrap  s

File: runtime/mangata-rococo/src/lib.rs [hunk 4]
```diff
@@ -474,6 +486,7 @@ impl pallet_xyk::Config for Runtime {
 	type DisabledTokens = Nothing;
 	type AssetMetadataMutation = AssetMetadataMutation;
 	type WeightInfo = weights::pallet_xyk_weights::ModuleWeight<Runtime>;
+	type RewardsMigrateAccount = RewardsMigrateAccountProvider<Self>;
```
Comment (mateuszaaa) [inline runtime/mangata-rococo/src/lib.rs:489]: please set same seession length && rewareds distribution period in rococo

**Review by mateuszaaa — APPROVED** [review]

File: pallets/issuance/src/lib.rs [hunk 5]
```diff
@@ -489,21 +528,31 @@ impl<T: Config> Pallet<T> {
...
+		let activated_pools: Vec<_> = PromotedPoolsRewardsV2::<T>::iter()
+			.filter_map(|(token_id, rewards)| {
+				T::ActivedPoolQueryApiType::get_pool_activate_amount(token_id)
+					.map(|activated_amount| (token_id, rewards, activated_amount))
+			})
+			.collect();
+		let mut liquidity_mining_issuance_per_pool = liquidity_mining_issuance;
+
+		if activated_pools.len() > 0 {
+			liquidity_mining_issuance_per_pool /= activated_pools.len() as u128;
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/issuance/src/lib.rs:544]: it is checked by: if activated_pools.len() > 0 , btw this was done by you:P

**Review by mateuszaaa — CHANGES_REQUESTED** [review]

File: runtime/mangata-rococo/src/lib.rs (hunk 4 above)
Comment (stanieltron) [reply runtime/mangata-rococo/src/lib.rs:489]: changed to be same

File: pallets/issuance/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/issuance/src/lib.rs:544]: it was more like some draft impl to help you proceed with benchmarking. AFAIK without that change benchmarks were failing, so we fixed it on screenshare session to cover some corner case ;)

File: pallets/xyk/src/lib.rs [hunk 6]
```diff
@@ -381,6 +403,28 @@ pub mod pallet {
...
+		CalculateCumulativeWorkMaxRatioMathError2,
+		CalculateCumulativeWorkMaxRatioMathError3,
+		CalculateCumulativeWorkMaxRatioMathError4,
+		CalculateCumulativeWorkMaxRatioMathError5,
+		CalculateCumulativeWorkMaxRatioMathError6,
+		CalculateCumulativeWorkMaxRatioMathError7,
+		CalculateRewardsAllMathError1,
+		CalculateRewardsAllMathError2,
+		CalculateRewardsAllMathError3,
+		NoRights,
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:427]: duplicates removed

File: pallets/xyk/src/lib.rs [hunk 7]
```diff
@@ -790,316 +889,247 @@ impl<T: Config> Pallet<T> {
...
-			last_checkpoint,
-			cummulative_work_in_last_checkpoint,
-			missing_at_last_checkpoint,
-		)
+			cummulative_work_max_possible_for_ratio =
+				liquidity_assets_amount_u256.checked_mul(U256::from(time_passed)).ok_or_else(
+					|| DispatchError::from(Error::<T>::CalculateCumulativeWorkMaxRatioMathError2),
+				)?;
+
+			// libm::floor(libm::pow(q, pow as f64) * REWARDS_PRECISION as f64) as u128
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:1002]: rmed

File: pallets/issuance/src/mock.rs [hunk 8]
```diff
@@ -31,6 +31,8 @@ use sp_runtime::{
 };
 use sp_std::convert::TryFrom;
 
+//use sp_std::sync::Mutex;
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/issuance/src/mock.rs:34]: removed

[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/issuance/src/mock.rs:34]: comment removed

File: pallets/xyk/src/lib.rs (hunk 2 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:904]: unwraps replaced by ok_or_else

File: pallets/xyk/src/lib.rs [hunk 9]
```diff
@@ -254,9 +254,20 @@ use sp_std::{
...
 
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct RewardInfo {
+	pub activated_amount: u128,
+	pub rewards_not_yet_claimed: u128,
+	pub rewards_already_claimed: u128,
+	pub last_checkpoint: u32,
+	pub pool_ratio_at_last_checkpoint: U256,
+	pub missing_at_last_checkpoint: U256,
+}
```
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/xyk/src/lib.rs:269]: still cant see the docs :thinking:

**Review by mateuszaaa — CHANGES_REQUESTED** [review]

**Review by mateuszaaa — APPROVED** [review]
lgtm

**Pull request merged**